src/librustc_passes/check_attr.rs

```diff
@@ -213,6 +223,99 @@ impl CheckAttrVisitor<'tcx> {
         }
     }
 
+    /// Checks if `#[cold]` is applied to a non-function. Returns `true` if valid.
+    fn check_cold(&self, attr: &Attribute, span: &Span, target: Target) -> bool {
+        match target {
+            Target::Fn | Target::Method(..) | Target::ForeignFn => true,
+            _ => {
```

Comment:
Does `#[cold]` work for expressions or statements?

Reply:
[The reference says no.](https://doc.rust-lang.org/reference/attributes/codegen.html#the-cold-attribute)

I think the equivalent for expressions is [`std::intrinsics::unlikely`](https://doc.rust-lang.org/std/intrinsics/fn.unlikely.html).

Reply:
I wouldn't use the reference as a source (speaking as the author of that content).

I'm pretty sure this is wrong for closures (although that needs `#![feature(stmt_expr_attributes)]`), and maybe some of these other checks.  

The codegen attributes are particularly hard to determine where they apply because you have to follow how they are interpreted inside llvm. 

Also, I think historically, adding checks for proper attribute locations uses the `unused_attributes` lint, due to the backwards compatibility hazard.

Reply:
Good point with closures--I think you're right.

Regarding backwards compatibility, I've seen a few different opinions while reading through the various issues.  I don't have a strong opinion but some of the arguments I've seen:
* Use `unused_attributes` for backwards compatibility
* Hard error is consistent with `repr`, `inline`, `target_feature`
* Hard error to avoid footguns like `#[link_name = "foo"] extern {}`
* Hard error to avoid ICEs such as #64768

Looking at what was done with `inline`, the lint was used to avoid breaking the bitflags crate: https://github.com/rust-lang/rust/pull/65294#issuecomment-545418714

I'm curious to see how many crates, if any, are broken by this change but I have no problem changing to a lint if that's necessary.